compiler/noirc_frontend/src/elaborator/enums.rs

```diff
@@ -243,4 +253,672 @@ impl Elaborator<'_> {
+        &mut self,
+        variable_to_match: DefinitionId,
+        rules: Vec<(Expression, Expression)>,
+    ) -> (Vec<Row>, Type) {
+        let result_type = self.interner.next_type_variable();
+        let expected_pattern_type = self.interner.definition_type(variable_to_match);
+
+        let rows = vecmap(rules, |(pattern, branch)| {
+            self.push_scope();
+            let pattern = self.expression_to_pattern(pattern, &expected_pattern_type);
+            let columns = vec![Column::new(variable_to_match, pattern)];
+
+            let guard = None;
+            let body_span = branch.span;
+            let (body, body_type) = self.elaborate_expression(branch);
+
+            self.unify(&body_type, &result_type, || TypeCheckError::TypeMismatch {
+                expected_typ: result_type.to_string(),
+                expr_typ: body_type.to_string(),
+                expr_span: body_span,
+            });
+
+            self.pop_scope();
+            Row::new(columns, guard, body)
+        });
+        (rows, result_type)
+    }
+
+    /// Convert an expression into a Pattern, defining any variables within.
+    fn expression_to_pattern(&mut self, expression: Expression, expected_type: &Type) -> Pattern {
+        let expr_span = expression.span;
+        let unify_with_expected_type = |this: &mut Self, actual| {
+            this.unify(actual, expected_type, || TypeCheckError::TypeMismatch {
+                expected_typ: expected_type.to_string(),
+                expr_typ: actual.to_string(),
+                expr_span,
+            });
+        };
+
+        match expression.kind {
+            ExpressionKind::Literal(Literal::Integer(x, negative)) => {
+                let Some(mut x): Option<i64> = x.try_to_u64().and_then(|x| x.try_into().ok())
+                else {
+                    panic!("integer too big! or small")
+                };
+                if negative {
+                    x = -x;
+                }
+
+                let actual = self.interner.next_type_variable_with_kind(Kind::IntegerOrField);
+                unify_with_expected_type(self, &actual);
+                Pattern::Int(x)
+            }
+            ExpressionKind::Variable(path) => {
+                // A variable can be free or bound if it refers to an enum constant:
+                // - in `(a, b)`, both variables may be free and should be defined, or
+                //   may refer to an enum variant named `a` or `b` in scope.
+                // - Possible diagnostics improvement: warn if `a` is defined as a variable
+                //   when there is a matching enum variant with name `Foo::a` which can
+                //   be imported. The user likely intended to reference the enum variant.
+                let path_len = path.segments.len();
+                let location = Location::new(path.span(), self.file);
+                let last_ident = path.last_ident();
+
+                match self.resolve_path(path) {
```

Comment:
Should probably use `resolve_path_or_error`. That will report any warnings, for example if a type was private but still exists.

(the name `resolve_path_or_error` is not ideal but I don't know what other name would be better)

Reply:
This is leftover from when I was calling `resolve_path` twice, calling it again in the helper function below and erroring only the second time. Seems I changed that but did not update here. I'll change it.